freqtrade/exchange/exchange.py

```diff
@@ -256,32 +256,21 @@ def set_sandbox(self, api: ccxt.Exchange, exchange_config: dict, name: str) -> N
                                      "Please check your config.json")
                 raise OperationalException(f'Exchange {name} does not provide a sandbox api')
 
-    def _load_async_markets(self, reload: bool = False) -> None:
-        try:
-            if self._api_async:
-                asyncio.get_event_loop().run_until_complete(
-                    self._api_async.load_markets(reload=reload))
-
-        except ccxt.BaseError as e:
-            logger.warning('Could not load async markets. Reason: %s', e)
-            return
-
```

Comment:
Removing this is wrong and can introduce unwanted side-effects. 

While we don't explicitly need this - ccxt will call this when doing anything using the async object  and it's not been explicitly called before.


So by not calling this at the start, we delay the first future operation. 
I'd rather have a slightly slower startup (i gues that's what you're trying to address here?) - and that's why this is there (might merit a comment explaining this ...)

Reply:
Well, as I see from the ccxt code (and this maybe found in the docs) -- it's a kind of convention in ccxt, for exchange methods to call `load_markets()` first if the method needs markets or market data for the certain market. And `load_market()` does absolutely same we do -- it fetches and processes markets data if it's yet empty and caches the previously fetched result if no `reload=True` was passed. And that's proper behavior -- to get all exchange methods idempotent, to call them in any order. We do not need to worry about this -- initialization should be properly handled and cached if needed in the intrinsics of ccxt. We also do not initialize currencies info and tickers, for example -- but you could also say "ccxt will call this when doing something", right? At least, "something ... ccxt is doing using the async object" here what we call for the ccxt async object and it's fetching ohlcv, and the fetch_ohlcv/fetch_ohlcvs methods (the default one, and the method from the particular exchange implementation) do always call `load_markets()` inside. Otherwise the exchange implementation will not pass the @kroitor verification, I guess...

If you think that some exchanges may not follow this, let's ask ccxt, should we initialize markets to fetch ohlcvs from the instance -- and if the answer will be "yes", my next question would be "what else should we initialize explicitly before fetching ohlcv?"... And if the next answer will be "only markets", my next question would be "why do we need to initialize markets and do not need to initialize other attributes -- currencies, tickers, whatever"...

I'm not addressing actually here slower startup (but this also would speedup the startup, yes) -- I'm removing strange dependencies, to make the code simplier and cleaner...

Reply:
^^ In short, in my opinion: it's not absolutely needed to call `load_markets` for async ccxt object if we do not use explicitly the `ccxt.market` attribute and assume it's initialized, if we only call the ccxt methods, that should initialize the markets inside, if they need it.

Reply:
well, i know that `load_markets()` is called within the ccxt methods - and i don't think there's differences in ccxt exchange implementations for this (let's assume all exchanges do this correctly).

It would still be a problem for freqtrade to not do this explicitly:

* we startup
* run the first loop to fetch data (this will be slower - as `load_markets()` will be called from within ccxt.
* ...

Now let's assume the worst case - kraken with a slightly older ccxt version (or uptodate version, but kraken (or any other exchange) changing something to break load_markets).
You would not notice anything during startup, but fetching candles will fail.

but not because fetch_ohlcv returns an error - but because load_markets() does not work correctly.

Now we introduced an implicit problem which needs to be handled in every exchange method - just because we don't want to call this explicitly.
Now the code did not become cleaner - but at lot more convoluted.


----

You're right - it's not absolutely needed as ccxt does this implicitly if it's not been called before.

To put it the way [the ccxt documentation puts it](https://github.com/ccxt/ccxt/wiki/Manual#loading-markets):

> If you forget to load markets the ccxt library will do that automatically upon your first call to the unified API. It will send two HTTP requests, first for markets and then the second one for other data, sequentially.

This implicitly states that you "forgot" to call it - which means that it's best practice to explicitly call this before.

Reply:
Okay, agree,

thanks for the link -- I was too lazy to find this part of the ccxt docs, sorry.

Reply:
however, if you don't mind - put a docstring to this method explaining this ... so we avoid this discussion in 3 months (or after a bottle of whisky / vodka for you) again :laughing:

Reply:
yeah, that's exactly what I was going to do with it, returning it back...